File: Algorithm/Stack/Solution/매치되는 괄호 찾기 - douv/match.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

int main(void)
{
    char str[51];
    scanf("%s", str);

    char stack[51][2];
    int size = 0;

    int pairs[25][2];
    int pairCount = 0;

    bool isMatched = true;

    for (int i = 0; i < strlen(str); ++i)
    {
        char ch = str[i];

        if (ch == '(' || ch == '{')
        {
            stack[size][0] = ch;
            stack[size][1] = i;
            size++;
        }
        else if (ch == ')')
        {
            if (size > 0 && stack[size - 1][0] == '(')
            {
                pairs[pairCount][0] = stack[size - 1][1];
                pairs[pairCount][1] = i;
                pairCount++;
                size--;
            }
            else
            {
                isMatched = false;
                break;
            }
        }
        else if (ch == '}')
        {
            if (size > 0 && stack[size - 1][0] == '{')
            {
                pairs[pairCount][0] = stack[size - 1][1];
                pairs[pairCount][1] = i;
                pairCount++;
                size--;
            }
            else
            {
                isMatched = false;
                break;
            }
        }
    }

    if (isMatched && size == 0)
    {
        for (int i = 0; i < pairCount; ++i)
        {
            printf("%d %d\n", pairs[i][0], pairs[i][1]);
        }
    }
    else
    {
        printf("not match\n");
    }

    return 0;
}

File: Algorithm/Stack/Solution/매치되는 괄호 찾기 - douv/match.py
s = input().strip()

stack = list()
pairs = list()
isMatched = True


for i, ch in enumerate(s):
    if ch == '(' or ch == '{':
        stack.append((ch, i))
    elif ch == ')':
        if len(stack) > 0 and stack[-1][0] == '(':
            pairs.append((stack[-1][1], i))
            stack.pop()
        else:
            isMatched = False
            break
    elif ch == '}':
        if len(stack) > 0 and stack[-1][0] == '{':
            pairs.append((stack[-1][1], i))
            stack.pop()
        else:
            isMatched = False
            break

if isMatched and len(stack) == 0:
    for pair in pairs:
        print("%d %d" % (pair[0], pair[1]))
else:
    print("not match")
